### PR TITLE
Add support for unix socket as upstream

### LINKED ISSUE #1865: Feature: support unix socket as upstream
## Expected Behavior

I'd like to be able to provide unix:///run/myapp/uvicorn.sock as the upstream server

~PR is coming~ see PR #1866 

## Your Environment

- Version used: 7.3.0

### PR DESCRIPTION
**Pull request by babs**
## Description

Add support for unix socket at upstream
Might require some rework.

Should be able to close #1865

## Motivation and Context

Some apps might require to go through a unix socket when listening on a tcp port, even bound locally is not secure enough and socket ownership is used as an extra isolation layer.

## How Has This Been Tested?

I created a simple uvicorn app that listen to unix socket and used `--upstream=unix:///path/pf/the/unix.sock` .

## Checklist:

- [x] My change requires a change to the documentation or CHANGELOG.
- [x] I have updated the documentation/CHANGELOG accordingly.
- [x] I have created a feature (non-master) branch for my PR.

### CODE REVIEW

File: pkg/upstream/http.go [hunk 1]
```diff
@@ -92,6 +95,19 @@ func (h *httpUpstreamProxy) ServeHTTP(rw http.ResponseWriter, req *http.Request)
 	}
 }
 
+type UnixRoundTripper struct {
+	Transport *http.Transport
+}
+
+func (t *UnixRoundTripper) RoundTrip(req *http.Request) (*http.Response, error) {
```
Comment (JoelSpeed) [inline pkg/upstream/http.go:102]: Please add comments to explain anything that is exported

File: pkg/upstream/http.go [hunk 2]
```diff
@@ -102,6 +118,15 @@ func newReverseProxy(target *url.URL, upstream options.Upstream, errorHandler Pr
 	// Inherit default transport options from Go's stdlib
 	transport := http.DefaultTransport.(*http.Transport).Clone()
 
+	if strings.HasPrefix(upstream.URI, "unix") {
```
Comment (JoelSpeed) [inline pkg/upstream/http.go:121]: Can we not use the scheme at this point rather than a string prefix? We have the target right?

**Review by JoelSpeed — COMMENTED** [review]
Would like to see some tests, similar to the HTTP tests, but we will need to work out how to create a temporary unix socket. I wonder if the HTTP test server supports this already? :thinking:

File: pkg/upstream/http.go (hunk 1 above)
Comment (babs) [reply pkg/upstream/http.go:102]: My bad, it shouldn't be exported, I'll change that.

File: pkg/upstream/http.go (hunk 2 above)
Comment (babs) [reply pkg/upstream/http.go:121]: I remember looking around specifically to avoid that but might have missed something

**Comment by babs** [comment]
How is currently started the upstream http to test against ?

File: pkg/upstream/http.go (hunk 1 above)
Comment (babs) [reply pkg/upstream/http.go]: Not exported anymore :)

File: pkg/upstream/http.go (hunk 2 above)
Comment (babs) [reply pkg/upstream/http.go]: Fixed also, I was looking mainly into upstream but the solution was in target :)

**Comment by JoelSpeed** [comment]
We create an [http testserver](https://github.com/oauth2-proxy/oauth2-proxy/blob/f753ec1ca5c42cd4c4e400afa1f76817638f3010/pkg/upstream/upstream_suite_test.go#L46-L48) in the suite setup, perhaps we can create something there that sets up a unix socket?

Otherwise code LGTM

Apologies for the delay in getting back to you

**Comment by babs** [comment]
I remember myself doing the server portion, but can't put my hand on it...

**Comment by babs** [comment]
still alive

**Comment by babs** [comment]
Unix mock is available at sock://test.sock
Help wanted to adapt http tests to add a pass with this mock as backend @JoelSpeed

**Comment by babs** [comment]
I miss the reminder of the bot.
Shouldn't be closed, still alive as far as I know.
Just need a little help to adapt tests to use the mock just like the regular http tests.

File: pkg/upstream/http.go [hunk 3]
```diff
@@ -92,6 +95,19 @@ func (h *httpUpstreamProxy) ServeHTTP(rw http.ResponseWriter, req *http.Request)
 	}
 }
 
+type unixRoundTripper struct {
```
Comment (JoelSpeed) [inline pkg/upstream/http.go:98]: Please add an appropriate godoc to the struct and fields within it

File: pkg/upstream/http.go [hunk 4]
```diff
@@ -92,6 +95,19 @@ func (h *httpUpstreamProxy) ServeHTTP(rw http.ResponseWriter, req *http.Request)
...
+	Transport *http.Transport
+}
+
+func (t *unixRoundTripper) RoundTrip(req *http.Request) (*http.Response, error) {
```
Comment (JoelSpeed) [inline pkg/upstream/http.go:102]: Missing godoc, would be good to explain how this differs from regular http transports

**Review by JoelSpeed — COMMENTED** [review]
Tests wise, lets make sure there's something added to the [proxy tests](https://github.com/oauth2-proxy/oauth2-proxy/blob/212359f467a77defc60e3feb0c7d2380d074c6df/pkg/upstream/proxy_test.go).

You'll need to add a new unix based upstream to the [upstreams](https://github.com/oauth2-proxy/oauth2-proxy/blob/212359f467a77defc60e3feb0c7d2380d074c6df/pkg/upstream/proxy_test.go#L44) and then add one or more [Entry](https://github.com/oauth2-proxy/oauth2-proxy/blob/212359f467a77defc60e3feb0c7d2380d074c6df/pkg/upstream/proxy_test.go#L141-L161) blocks that exercises that new endpoint

File: pkg/upstream/http.go (hunk 3 above)
Comment (babs) [reply pkg/upstream/http.go:102]: done :)

File: pkg/upstream/http.go (hunk 4 above)
Comment (babs) [reply pkg/upstream/http.go:107]: done :)

**Comment by babs** [comment]
Review needed still relevant

**Comment by babs** [comment]
damn, I missed the reminder @JoelSpeed :/

**Comment by babs** [comment]
Knowing that test have been added, is there anything else to do to see this PR merged ?

**Comment by tuunit** [comment]
@babs I'll try to find some time to test it on the weekend :)

File: CHANGELOG.md [hunk 5]
```diff
@@ -43,6 +43,7 @@ to remain consistent with CLI flags. You should specify `code_challenge_method`
 - [#1851](https://github.com/oauth2-proxy/oauth2-proxy/pull/1851) Bump golang to 1.19 and min allowed version to 1.18
 - [#1815](https://github.com/oauth2-proxy/oauth2-proxy/pull/1815) Keycloak: save user and preferredUsername in session to populate headers for the backend (@babs)
 - [#1847](https://github.com/oauth2-proxy/oauth2-proxy/pull/1847) Update go-redis/redis to v9 (@arhamGH)
+- [#1866](https://github.com/oauth2-proxy/oauth2-proxy/pull/1866) Add support for unix socker as upstream (@babs)
```
Comment (JoelSpeed) [inline CHANGELOG.md:46]: This needs to be moved to the correct place in the changelog now, otherwise will get this merged

Comment (babs) [reply CHANGELOG.md]: Rebased and changelog updated accordingly;)

**Review by JoelSpeed — APPROVED** [review]
Thanks @babs

**Pull request merged**